### ISSUE TITLE
Hi, how to make it render the PDF not just as Image, but as HTML like pdf.js do?

### ISSUE THREAD

**Issue by hz0324** [nnarhinen/react-pdf#52]
Thanks.

**Comment by wojtekmaj** [comment]
Hi,
It's the first time I hear about PDF.js rendering HTML instead of canvas element which I use here. There are PDF.js based full sized readers that could do that and maybe one day I'd device to make one too as I see there are quite a bit of requests for such, with ability to select text, have all pages rendered at the same time etc., but so far I'm unaware of React-based solution for this. Or I'm terribly wrong, in that case, could you show me an example?

**Comment by oyeanuj** [comment]
@wojtekmaj Also came here looking for the same. So far, haven't come across a React-based solution but the main demo on `pdf.js` shows a HTML based version - https://mozilla.github.io/pdf.js/web/viewer.html

Is it possible to incorporate the viewer as `dangerouslySetInnerHTML` or does that mean a rewrite of the renderer?

**Comment by hz0324** [comment]
@oyeanuj Thanks for the information. I thought it maybe possible to include the whole pdf.js demo to a iframe inside one React component. And the changing the demo code to fit our needs. I haven't tried this yet, but seems like a possible solution. 

Please let me know, if you have any other thoughts. Thanks.

**Comment by hz0324** [comment]
@wojtekmaj oyeanuj has pointed the online demo. That's what we want, but I really don't know how to do it due to the lack of documentation of PDF.js.

If you can do it, please help us. Thanks.

**Comment by wojtekmaj** [comment]
@oyeanuj, @hz0324

TL;DR There are legal and technical problems now, I have plans for PDF reader (which React-PDF is not) in the near future after React-PDF 2.0 is released.

I'm aware of this demo. It's based on HTML, of course, but the pages themselves are not HTML, they are still rendered in canvas.

I'd like to stress once again that building a full sized solution like this is way, way more complex than React-PDF. It involves things like navigation, multi page rendering, ability to search and select text, automatic scaling and much much more.

Of course you could go the lazy way and make an iframe with pdfjs-based reader like this (Mozilla made a lot of confusion with naming here...). There are some problems with this approach though. First of all, Mozilla is not okay with using it as-is. Will cite when I'm not on the go. Secondly - it would still be an iframe which we would have limited control over and thus I have no plans delivering such half-baked solution.

There's good news though. While I'm extremely busy in my private and work life, it looks like React-PDF 2.0 is slowly approaching. I have already planned its architecture with creating fully fledged PDF readers in mind. The next natural step would be to try and build one!

Hope this helps!

**Comment by hz0324** [comment]
@wojtekmaj thanks for the reply. 

I am curious about your plan to build a fully fledged PDF reader. Are you going to write the render engine yourself. Or using an existing one, such as PDF.js's? 
As far as I thought, building an engine is a huge project, while building a reader view is much easier/

**Comment by wojtekmaj** [comment]
@hz0324 Definitely I'm not going to re-write the renderer.

What I have now is an ability to render a single page on a <canvas> element. My goal is to use this project to create another one, and create an interface that would display all the pages in the PDF along with view & navigation features, like thumbnails, ToC etc.

**Comment by hz0324** [comment]
@wojtekmaj  Got it. Good luck.

I know PDF.js has a feature to layout the text in P tag and their coordination info, I have seen this on other people's react related PDF project. But I cannot found it. You may use it to make selectable.

**Comment by oyeanuj** [comment]
@wojtekmaj Thank you for that update, very helpful!

@hz0324 I think you might be referring to this - https://github.com/mozilla/pdf.js/issues/7072 ?

**Comment by hz0324** [comment]
@oyeanuj Yes, this is what I am talking about. But now, we have decided to use PDF.js default viewer inside an iframe of a react component. We will modify the viewer's css to meet our need. Haven't seen much problem now. Hope it works.

Good luck to your project.

**Comment by oyeanuj** [comment]
@hz0324 Keep us posted on what solution you end up with!

**Comment by wojtekmaj** [comment]
Hey @oyeanuj and @hz0324!
I'm super excited to announce that React-PDF v2.0.0-alpha.2 now supports text layers!

Check this out on alpha demo!
http://projekty.wojtekmaj.pl/react-pdf/alpha/

I'd love to hear what you think :)

**Comment by oyeanuj** [comment]
@wojtekmaj Very cool! Couple of things that came to mind - 

1. It doesn't seem to support text selection, copy or paste at the moment. Not sure if you already knew about it but wanted to bring it up.

2. For larger documents, have you considered something like React-Virtualized, so that the load on the browser is manageable?

**Comment by wojtekmaj** [comment]
@oyeanuj 

Ad 1. Hate to be that guy, but "Works on my machine" :D 

![image](https://user-images.githubusercontent.com/5426427/28253402-fa642bba-6aa3-11e7-8335-15fcd13ffb5c.png)

What have you tried? What browser are you testing on? I'd love to help but can't reproduce :(

Ad 2. Yes, I do! That's my plan for react-pdf-reader, which I will be FINALLY able to create based on react-pdf 2.0 thanks to its much, much better architecture :)

**Comment by oyeanuj** [comment]
@wojtekmaj Ha, I see something different. Screenshot below, on Chrome 60 (Beta Build).

<img width="1474" alt="screen shot 2017-07-16 at 7 07 10 pm" src="https://user-images.githubusercontent.com/9633371/28253567-1b63db0c-6a5a-11e7-9aef-5d6a116ffa7d.png">

Looking forward to the react-pdf-reader!

**Comment by wojtekmaj** [comment]
Hah, that's my new renderer going crazy on HiDPI screens. Will fix that soon :) Thanks!

Edit: Fixed! Also, it turns out that React-PDF's textLayers are much more accurate than Mozilla's own ones :)

![image](https://user-images.githubusercontent.com/5426427/28254258-7889bc9c-6aab-11e7-9cc1-9cbe324cc700.png)
